### ISSUE TITLE
"Latest" feed occasionally shows posts out of order

### ISSUE THREAD

**Issue by aitchiss** [forem/forem#14709]
**Describe the bug**

Occasionally, when viewing the `/latest` feed, users can see more recently published posts appearing below older posts. For example:

![screenshot of DEV latest feed. A post is circled that shows it was published "1 hour ago", and above and below it are posts from 2 hours ago](https://user-images.githubusercontent.com/20773163/133114102-b93cac35-5da8-4f47-8f02-3c15fbbc3930.png)

Initially I thought this was an issue coming from the API, but I think it might actually be to do with [how the frontend constructs and inserts paged-in articles in the feed](https://github.com/forem/forem/blob/16cec7d93d04048947f64c14e40c46ce9cd337d1/app/assets/javascripts/initializers/initScrolling.js#L218-L289). In the example screengrab above, the network responses shows that the "Tracking responses..." article should have appeared in a different part of the sequence (the surrounding articles pictured above were not returned from the same network fetch).

TL;DR - it needs more investigation, but I think the source of the issue may be in how the paging behaviour plays with the chronological feed.

**To Reproduce**

- Go to the `/latest` feed
- Start scrolling and keep an eye on the published at date on the feed cards
- Eventually you'll see some that appear out of chronological order

**Expected behavior**

The articles in the "latest" feed should appear in the order they were published at, with the newest first.

**Comment by djuber** [comment]
It looks like a way to replicate this would be
- load /latest - do not scroll down
- wait for new articles to be published (20-30 minutes is probably enough for DEV.to, basically wait for new articles to have been published) - in test you can create articles to suit your needs.
- scroll down until a request to /search/feed_content?... is made (which returns a json list of articles)
- observe the placement of the new items in the existing list (you can view the search response and text search by title on the page) 

I'm definitely able to replicate this behavior (new items appear out of order in the latest feed when scrolling). 

I think there are a few strategies that could be followed, I'm not sure any of them are obviously correct without drawbacks

- backfilling articles items to their appropriate location in the list when they're loaded from the paged request (if an article is earlier than the displayed region - it's inserted in the page you've already scrolled past) - this could cause new articles to be placed out of view (it's unlikely someone would scroll backward looking for changes often)
- sending a "earlier than" mark based on the last article (oldest we already have) rather than a page number - this would prevent newer items (published since the page was loaded) from appearing at all, but should be easier to keep items in order.
- accept the current state of things, that sometimes this gives a confusing result in what looked like a sorted list

**Comment by aitchiss** [comment]
I think there might be two aspects to this bug. The one that you mention @djuber, but I think there's also something else funky going on... As far as I know, the expectation is that the UI should present posts in the order they're received in the API response, but once we start scrolling, that isn't always happening.

You can reproduce in DEV fairly reliably by

- visiting /latest
- immediately start scrolling
- eventually (for me it's after I start hitting posts from 2 hours ago-ish) you'll see something like the screenshot in the description above
- when I inspect the network requests the "out of place" article was actually sent from the API to appear between posts of a similar published time. For some reason it's being inserted into the UI in a much later position.

I'd need to investigate further but my initial suspicion is there might be a timing issue occurring where multiple pages of posts are loaded in quick succession and the individual callbacks to update the UI are overlapping each other.

**Comment by pramodrhegde** [comment]
@aitchiss I would love to make a contribution here. please assign this to me. thanks!

**Comment by pramodrhegde** [comment]
@cmgorton @aitchiss I would love to work on this issue, could you please assign this to me? thanks!

**Comment by madzadev** [comment]
Yup, tested today myself by creating a new post and the issue still exists.
Below is the GIF where the Latest feed position my article that has been posted 2 mins before.

![RYbrcKwBZZ](https://user-images.githubusercontent.com/42511917/137714472-a11aed85-227a-4a2b-b93e-3efb1e19ad37.gif)

Some other articles are out of order as well.
Take a look @benhalpern , I think this might be a priority.

**Comment by pramodrhegde** [comment]
> I was going to report this issue, I'm glad someone already posted. Any updates on this @pramodrhegde?

Apologies for the delayed response, I didn't get a chance to look into this, I am unassigning myself so that anyone else can take this up.

**Comment by jeremyf** [comment]
There's an open pull request to experiment with a different feed algorithm: https://github.com/forem/forem/pull/15240, but there remains a particular challenge.  Namely sorting across pagination pages.

In Forem there are 2 possible feed strategies:  [Basic](https://github.com/forem/forem/blob/6b2160f307b41467bcf4c0f63eebccaa5a5bd341/app/services/articles/feeds/basic.rb) and [LargeForemExperimental](https://github.com/forem/forem/blob/6b2160f307b41467bcf4c0f63eebccaa5a5bd341/app/services/articles/feeds/large_forem_experimental.rb).  Each has (at present) has a different pagination size.  The Basic is 25 and the other is 50.

DEV.to uses LargeForemExperimental.  So, when you scroll to the 51st article, you're now seeing a new pagination window, and that sorting "resets".

Regarding the `/latest` feed, I'm uncertain of the why of that particular issue, but we might be encountering timing issues.  In the `/latest` feed we're sorting reverse publication date.  We are also likely caching those results, which in turn can create out of sequence entries.

**Comment by jeremyf** [comment]
To address both pagination and caching, it's possible we would need to introduce a `AND publication_date <= cached.request.time`. That would help ensure we are not "out of order" but would instead hide newly published articles until the cache expired.

tl;dr : We need to tease apart expectations of chronology for the "default" feed and for the "latest" feed as it relates to pagination, caching, and user interests.

**Comment by madzadev** [comment]
@pramodrhegde no worries, it's all good!

**Comment by jeremyf** [comment]
@madzadev The current work that I'm looking to roll-out is an experiment (if you're Rails inclined, [here's the PR](https://github.com/forem/forem/pull/15240)).  As written, the experiment will impact 20% of the people (alas I don't have a lever to conscript you in that experiment).

**Comment by madzadev** [comment]
@jeremyf Curious have there been any experiments related to feed behavior on specific percentage of users before? 
It seems exactly like that, cause all the indexing issues in the feed happen just for the specific group of people. For the most of users the feed (both latest and main) still works fine, as expected.

**Comment by ephaeton** [comment]
![image](https://github.com/forem/forem/assets/406393/9528060f-a125-4b86-9975-d3d08a8df546)

no scrolling, just clicked latest.
alternatively, (re-)visit URL /latest

As you can see, the first post is older than the second displayed, which, AFAIUI, shouldn't be the case.
